# Homework and bake-off: Sentiment analysis

In [ ]:
__author__ = "Christopher Potts"
__version__ = "CS224u, Stanford, Spring 2021"

## Contents

1. [Overview](#Overview)
1. [Methodological note](#Methodological-note)
1. [Set-up](#Set-up)
1. [Train set](#Train-set)
1. [Dev sets](#Dev-sets)
1. [A softmax baseline](#A-softmax-baseline)
1. [RNNClassifier wrapper](#RNNClassifier-wrapper)
1. [Error analysis](#Error-analysis)
1. [Homework questions](#Homework-questions)
  1. [Token-level differences [1 point]](#Token-level-differences-[1-point])
  1. [Training on some of the bakeoff data [1 point]](#Training-on-some-of-the-bakeoff-data-[1-point])
  1. [A more powerful vector-averaging baseline [2 points]](#A-more-powerful-vector-averaging-baseline-[2-points])
  1. [BERT encoding [2 points]](#BERT-encoding-[2-points])
  1. [Your original system [3 points]](#Your-original-system-[3-points])
1. [Bakeoff [1 point]](#Bakeoff-[1-point])
1. [Submission Instruction](#Submission-Instruction)

## Overview

This homework and associated bakeoff are devoted to supervised sentiment analysis using the ternary (positive/negative/neutral) version of the Stanford Sentiment Treebank (SST-3) as well as a new dev/test dataset drawn from restaurant reviews. Our goal in introducing the new dataset is to push you to create a system that performs well in both the movie and restaurant domains.

The homework questions ask you to implement some baseline system, and the bakeoff challenge is to define a system that does well at both the SST-3 test set and the new restaurant test set. Both are ternary tasks, and our central bakeoff score is the mean of the macro-FI scores for the two datasets. This assigns equal weight to all classes and datasets regardless of size.

The SST-3 test set will be used for the bakeoff evaluation. This dataset is already publicly distributed, so we are counting on people not to cheat by developing their models on the test set. You must do all your development without using the test set at all, and then evaluate exactly once on the test set and turn in the results, with no further system tuning or additional runs. __Much of the scientific integrity of our field depends on people adhering to this honor code__. 

One of our goals for this homework and bakeoff is to encourage you to engage in __the basic development cycle for supervised models__, in which you

1. Design a new system. We recommend starting with something simple.
1. Use `sst.experiment` to evaluate your system, using random train/test splits initially.
1. If you have time, compare your system with others using `sst.compare_models` or `utils.mcnemar`. (For discussion, see [this notebook section](sst_02_hand_built_features.ipynb#Statistical-comparison-of-classifier-models).)
1. Return to step 1, or stop the cycle and conduct a more rigorous evaluation with hyperparameter tuning and assessment on the `dev` set.

[Error analysis](#Error-analysis) is one of the most important methods for steadily improving a system, as it facilitates a kind of human-powered hill-climbing on your ultimate objective. Often, it takes a careful human analyst just a few examples to spot a major pattern that can lead to a beneficial change to the feature representations.

## Methodological note

You don't have to use the experimental framework defined below (based on `sst`). The only constraint we need to place on your system is that it must have a `predict_one` method that can map directly from an example text to a prediction, and it must be able to make predictions without having any information beyond the text. (For example, it can't depend on knowing which task the text comes from.) See [the bakeoff section below](#Bakeoff-[1-point]) for examples of functions that conform to this specification.

## Set-up

See [the first notebook in this unit](sst_01_overview.ipynb#Set-up) for set-up instructions.

In [ ]:
from collections import Counter
import numpy as np
import os
import pandas as pd
from sklearn.linear_model import LogisticRegression
import torch.nn as nn

from torch_rnn_classifier import TorchRNNClassifier
from torch_tree_nn import TorchTreeNN
import sst
import utils

In [ ]:
SST_HOME = os.path.join('data', 'sentiment')

## Train set

Our primary train set is the SST-3 train set:

In [ ]:
sst_train = sst.train_reader(SST_HOME)

In [ ]:
sst_train.shape[0]

8544

This is the train set we will use for all the regular homework questions. You are welcome to bring in new datasets for your original system. You are also free to add `include_subtrees=True`. This is very likely to lead to better systems, but it substantially increases the overall size of the dataset (from 8,544 examples to 159,274), which will in turn substantially increase the time it takes to run experiments.

See [this notebook](sst_01_overview.ipynb) for additional details of this dataset.

## Dev sets

We have two development set. SST3-dev consists of sentences from movie reviews, just like SST-3 train:

In [ ]:
sst_dev = sst.dev_reader(SST_HOME)

In [ ]:
sst_dev.shape

(1101, 4)

Our new bakeoff dev set consists of sentences from restaurant reviews:

In [ ]:
bakeoff_dev = sst.bakeoff_dev_reader(SST_HOME)

In [ ]:
bakeoff_dev.sample(3, random_state=1).to_dict(orient='records')

[{'example_id': 57,
  'is_subtree': 0,
  'label': 'neutral',
  'sentence': 'I would recommend that you make reservations in advance.'},
 {'example_id': 590,
  'is_subtree': 0,
  'label': 'positive',
  'sentence': 'We were welcomed warmly.'},
 {'example_id': 1968,
  'is_subtree': 0,
  'label': 'neutral',
  'sentence': 'We have been to Oceanaire twice in the last 6 weeks.'}]

Here is the label distribution:

In [ ]:
bakeoff_dev.label.value_counts()

neutral     1019
positive     777
negative     565
Name: label, dtype: int64

In [ ]:
bakeoff_dev.shape

(2361, 4)

The label distribution for the corresponding test set is similar to this.

## A softmax baseline

This example is here mainly as a reminder of how to use our experimental framework with linear models:

In [ ]:
def unigrams_phi(text):
    return Counter(text.split())

Thin wrapper around `LogisticRegression` for the sake of `sst.experiment`:

In [ ]:
def fit_softmax_classifier(X, y):
    mod = LogisticRegression(
        fit_intercept=True,
        solver='liblinear',
        multi_class='ovr')
    mod.fit(X, y)
    return mod

The experimental run with some notes:

In [ ]:
softmax_experiment = sst.experiment(
    sst.train_reader(SST_HOME),   # Train on any data you like except SST-3 test!
    unigrams_phi,                 # Free to write your own!
    fit_softmax_classifier,       # Free to write your own!
    assess_dataframes=[sst_dev, bakeoff_dev]) # Free to change this during development!

Assessment dataset 1
              precision    recall  f1-score   support

    negative      0.628     0.689     0.657       428
     neutral      0.343     0.153     0.211       229
    positive      0.629     0.750     0.684       444

    accuracy                          0.602      1101
   macro avg      0.533     0.531     0.518      1101
weighted avg      0.569     0.602     0.575      1101

Assessment dataset 2
              precision    recall  f1-score   support

    negative      0.272     0.690     0.390       565
     neutral      0.429     0.113     0.179      1019
    positive      0.409     0.346     0.375       777

    accuracy                          0.328      2361
   macro avg      0.370     0.383     0.315      2361
weighted avg      0.385     0.328     0.294      2361

Mean of macro-F1 scores: 0.416


`softmax_experiment` contains a lot of information that you can use for error analysis; see [this section below](#Error-analysis) for starter code.

## RNNClassifier wrapper

This section illustrates how to use `sst.experiment` with `TorchRNNClassifier`.

To featurize examples for an RNN, we can just get the words in order, letting the model take care of mapping them into an embedding space.

In [ ]:
def rnn_phi(text):
    return text.split()

The model wrapper gets the vocabulary using `sst.get_vocab`. If you want to use pretrained word representations in here, then you can have `fit_rnn_classifier` build that space too; see [this notebook section for details](sst_03_neural_networks.ipynb#Pretrained-embeddings). See also [torch_model_base.py](torch_model_base.py) for details on the many optimization parameters that `TorchRNNClassifier` accepts.

In [ ]:
def fit_rnn_classifier(X, y):
    sst_glove_vocab = utils.get_vocab(X, mincount=2)
    mod = TorchRNNClassifier(
        sst_glove_vocab,
        early_stopping=True)
    mod.fit(X, y)
    return mod

In [ ]:
rnn_experiment = sst.experiment(
    sst.train_reader(SST_HOME),
    rnn_phi,
    fit_rnn_classifier,
    vectorize=False,  # For deep learning, use `vectorize=False`.
    assess_dataframes=[sst_dev, bakeoff_dev])

Stopping after epoch 41. Validation score did not improve by tol=1e-05 for more than 10 epochs. Final error is 1.6700571030378342

Assessment dataset 1
              precision    recall  f1-score   support

    negative      0.592     0.617     0.604       428
     neutral      0.261     0.214     0.235       229
    positive      0.636     0.669     0.652       444

    accuracy                          0.554      1101
   macro avg      0.496     0.500     0.497      1101
weighted avg      0.541     0.554     0.547      1101

Assessment dataset 2
              precision    recall  f1-score   support

    negative      0.253     0.625     0.360       565
     neutral      0.431     0.246     0.314      1019
    positive      0.439     0.218     0.291       777

    accuracy                          0.327      2361
   macro avg      0.374     0.363     0.322      2361
weighted avg      0.391     0.327     0.317      2361

Mean of macro-F1 scores: 0.409


## Error analysis

This section begins to build an error-analysis framework using the dicts returned by `sst.experiment`. These have the following structure:

```
'model': trained model
'phi': the feature function used
'train_dataset':
   'X': feature matrix
   'y': list of labels
   'vectorizer': DictVectorizer,
   'raw_examples': list of raw inputs, before featurizing   
'assess_datasets': list of datasets, each with the same structure as the value of 'train_dataset'
'predictions': list of lists of predictions on the assessment datasets
'metric': `score_func.__name__`, where `score_func` is an `sst.experiment` argument
'score': the `score_func` score on the each of the assessment dataasets
```
The following function just finds mistakes, and returns a `pd.DataFrame` for easy subsequent processing:

In [ ]:
def find_errors(experiment):
    """Find mistaken predictions.

    Parameters
    ----------
    experiment : dict
        As returned by `sst.experiment`.

    Returns
    -------
    pd.DataFrame

    """
    dfs = []
    for i, dataset in enumerate(experiment['assess_datasets']):
        df = pd.DataFrame({
            'raw_examples': dataset['raw_examples'],
            'predicted': experiment['predictions'][i],
            'gold': dataset['y']})
        df['correct'] = df['predicted'] == df['gold']
        df['dataset'] = i  # two types of assess datasets (that will be conatenated into one)
        dfs.append(df)
    return pd.concat(dfs)

In [ ]:
softmax_analysis = find_errors(softmax_experiment)

In [ ]:
rnn_analysis = find_errors(rnn_experiment)

Here we merge the sotmax and RNN experiments into a single DataFrame:

In [ ]:
# if name collision (col_i -> col_i_x (left), col_i_y (right))
analysis = softmax_analysis.merge(
    rnn_analysis, left_on='raw_examples', right_on='raw_examples')
# remove excess 'gold' column 
analysis = analysis.drop('gold_y', axis=1).rename(columns={'gold_x': 'gold'})  

In [ ]:
analysis.head()

raw_examples  ... dataset_y
0  It 's a lovely film with lovely performances b...  ...         0
1  No one goes unindicted here , which is probabl...  ...         0
2  And if you 're not nearly moved to tears by a ...  ...         0
3                   A warm , funny , engaging film .  ...         0
4  Uses sharp humor and insight into human nature...  ...         0

[5 rows x 8 columns]

In [ ]:
# dataset_x and dataset_y are identical
analysis['dataset_x'].equals(analysis['dataset_y'])

True

The following code collects a specific subset of examples; small modifications to its structure will give you different interesting subsets:

In [ ]:
# Examples where the softmax model is correct, the RNN is not,
# and the gold label is 'positive'

error_group = analysis[
    (analysis['predicted_x'] == analysis['gold'])
    &
    (analysis['predicted_y'] != analysis['gold'])
    &
    (analysis['gold'] == 'positive')
]

In [ ]:
error_group.shape[0]

223

In [ ]:
for ex in error_group['raw_examples'].sample(5, random_state=1):
    print("="*70)
    print(ex)

We made online reservations and we seating exactly at our requested time.
That is a compliment to Kuras and Miller .
We had eaten dinner at Caffe Molise and enjoyed the food, deciding to come back.
Based on a devilishly witty script by Heather McGowan and Niels Mueller , the film gets great laughs , but never at the expense of its characters
A perfect spot for Business or special events.


## Homework questions

Please embed your homework responses in this notebook, and do not delete any cells from the notebook. (You are free to add as many cells as you like as part of your responses.)

### Token-level differences [1 point]

We can begin to get a sense for how our two dev sets differ by considering the most frequent tokens from each. This question asks you to begin such analysis.

Your task: write a function `get_token_counts` that, given a `pd.DataFrame` in the format of our datasets, tokenizes the example sentences based on whitespace and creates a count distribution over all of the tokens. The function should return a `pd.Series` sorted by frequency; if you create a count dictionary `d`, then `pd.Series(d).sort_values(ascending=False)` will give you what you need.

In [ ]:
def get_token_counts(df):
    """Count tokens frequency of a dataset, 
    where a sentence is a list of tokens separated by whitespaces

    Parameters
    ----------
    df : pd.DataFrame
        A pd.Dataframe with 'sentence' column, where each cell is a sentence (string of words)

    Returns
    -------
    pd.Series
        A sorted word counts pd.Series in descending order
    """
    
    ws_sentence_split = df.sentence.apply(lambda sentence: sentence.split())
    d = Counter(ws_sentence_split.explode())
    return pd.Series(d).sort_values(ascending=False)



In [ ]:
def test_get_token_counts(func):
    df = pd.DataFrame([
        {'sentence': 'a a b'},
        {'sentence': 'a b a'},
        {'sentence': 'a a a b.'}])
    result = func(df)
    for token, expected in (('a', 7), ('b', 2), ('b.', 1)):
        actual = result.loc[token]
        assert actual == expected, \
            "For token {}, expected {}; got {}".format(
            token, expected, actual)

In [ ]:
if 'IS_GRADESCOPE_ENV' not in os.environ:
    test_get_token_counts(get_token_counts)

In [ ]:
dev_token_counts = get_token_counts(sst_dev)
bakeoff_dev_token_counts = get_token_counts(bakeoff_dev)

In [ ]:
dev_token_counts

.         1037
,          935
the        810
of         585
and        576
          ... 
buyer        1
tube         1
lesser       1
Ford         1
Land         1
Length: 5374, dtype: int64

In [ ]:
bakeoff_dev_token_counts

the              1402
a                 886
and               818
to                750
was               731
                 ... 
constricting.       1
Tues-Fri            1
Barnsboro           1
$300.00             1
discover            1
Length: 5846, dtype: int64

As you develop your original system, you might review these results. The two dev sets have different vocabularies and different low-level encoding details that are sure to impact model performance, especially when one considers that the train set is like `sst_dev` in all these respects. For additional discussion, see [this notebook section](sst_01_overview.ipynb#Tokenization).

### Training on some of the bakeoff data [1 point]

We have so far presented the bakeoff dev set as purely for evaluation. Since the train set consists entirely of SST-3 data, this makes the bakeoff split especially challenging. We might be able to reduce the challenging by adding some of the bakeoff dev set to the train set, keeping some of it for evaluation. The current question asks to begin explore the effects of such training.

Your task: write a function `run_mixed_training_experiment`. The function should:

1. Take as inputs (a) a model training wrapper like `fit_softmax_classifier` and (b) an integer `bakeoff_train_size` specifying the number of examples from `bakeoff_dev` that should be included in the train set.
1. Split `bakeoff_dev` so that the first `bakeoff_train_size` examples are in the train set and the rest are used for evaluation.
1. Use `sst.experiment` with the user-supplied model training wrapper, `unigram_phi` as defined above, and a train set that consists of SST-3 train and the train portion of `bakeoff_dev` as defined in step 2. The value of `assess_dataframes` should be a list consisting of the SST-3 dev set and the evaluation portion of `bakeoff_dev` as defined in step 2.
1. Return the return value of `sst.experiment`.

The function `test_run_mixed_training_experiment` will help you iterate to the required design.

In [ ]:
sst_train.head() 

example_id  ... is_subtree
0    00001-00001  ...          0
71   00002-00001  ...          0
144  00003-00001  ...          0
221  00004-00001  ...          0
258  00005-00001  ...          0

[5 rows x 4 columns]

In [ ]:
def run_mixed_training_experiment(wrapper_func, bakeoff_train_size):
    """Create an experiment where the training includes additional bakeoff_train_size examples 
    from bakeoff development set

    Parameters
    ----------
    wrapper_func : function
        a wrapper function to create a model fit

    bakeoff_train_size: int
        the amount of examples from bakeoff_dev to be on training set

    Returns
    -------
    dictionary
        an experiment dictionary
    """
    mixed_training_experiment = sst.experiment(
        pd.concat((sst.train_reader(SST_HOME), bakeoff_dev.iloc[:bakeoff_train_size])),
        unigrams_phi,
        wrapper_func,
        assess_dataframes=[sst_dev, bakeoff_dev.iloc[bakeoff_train_size:]]
        )
    return mixed_training_experiment

In [ ]:
def test_run_mixed_training_experiment(func):
    bakeoff_train_size = 1000
    experiment = func(fit_softmax_classifier, bakeoff_train_size)

    assess_size = len(experiment['assess_datasets'])
    assert len(experiment['assess_datasets']) == 2, \
        ("The evaluation should be done on two datasets: "
         "SST3 and part of the bakeoff dev set. "
         "You have {} datasets.".format(assess_size))

    bakeoff_test_size = bakeoff_dev.shape[0] - bakeoff_train_size
    expected_eval_examples = bakeoff_test_size + sst_dev.shape[0]
    eval_examples = sum(len(d['raw_examples']) for d in experiment['assess_datasets'])
    assert expected_eval_examples == eval_examples, \
        "Expected {} evaluation examples; got {}".format(
        expected_eval_examples, eval_examples)

In [ ]:
if 'IS_GRADESCOPE_ENV' not in os.environ:
    test_run_mixed_training_experiment(run_mixed_training_experiment)

Assessment dataset 1
              precision    recall  f1-score   support

    negative      0.627     0.671     0.648       428
     neutral      0.319     0.162     0.214       229
    positive      0.638     0.757     0.692       444

    accuracy                          0.599      1101
   macro avg      0.528     0.530     0.518      1101
weighted avg      0.567     0.599     0.576      1101

Assessment dataset 2
              precision    recall  f1-score   support

    negative      0.471     0.412     0.440       320
     neutral      0.588     0.590     0.589       612
    positive      0.503     0.548     0.525       429

    accuracy                          0.535      1361
   macro avg      0.521     0.517     0.518      1361
weighted avg      0.534     0.535     0.534      1361

Mean of macro-F1 scores: 0.518


### A more powerful vector-averaging baseline [2 points]

In [Distributed representations as features](sst_03_neural_networks.ipynb#Distributed-representations-as-features), we looked at a baseline for the ternary SST-3 problem in which each example is modeled as the mean of its GloVe representations. A `LogisticRegression` model was used for prediction. A neural network might do better with these representations, since there might be complex relationships between the input feature dimensions that a linear classifier can't learn. To address this question, we want to get set up to run the experiment with a shallow neural classifier. 

Your task: write and submit a model wrapper function around `TorchShallowNeuralClassifier`. This function should implement hyperparameter search according to this specification:

* Set `early_stopping=True` for all experiments.
* Using 3-fold cross-validation, exhaustively explore this set of hyperparameter combinations:
  * The hidden dimensionality at 50, 100, and 200.
  * The hidden activation function as `nn.Tanh()` and `nn.ReLU()`.
* For all other parameters to `TorchShallowNeuralClassifier`, use the defaults.

See [this notebook section](sst_02_hand_built_features.ipynb#Hyperparameter-search) for examples. You are not required to run a full evaluation with this function using `sst.experiment`, but we assume you will want to.

We're not evaluating the quality of your model. (We've specified the protocols completely, but there will still be variation in the results.) However, the primary goal of this question is to get you thinking more about this strong baseline feature representation scheme for SST-3, so we're sort of hoping you feel compelled to try out variations on your own.

In [ ]:
from torch_shallow_neural_classifier import TorchShallowNeuralClassifier

def fit_shallow_neural_classifier_with_hyperparameter_search(X, y):
    
    basemod = TorchShallowNeuralClassifier(early_stopping=True)  # max_iter=10, n_iter_no_change=5
    cv = 3
    param_grid = {'hidden_dim': [50, 100, 200], 'hidden_activation': [nn.Tanh(), nn.ReLU()]}
    
    bestmod = utils.fit_classifier_with_hyperparameter_search(
        X, y, basemod, cv=cv, param_grid=param_grid)  # scoring='f1_macro'
    return bestmod


In [ ]:
bakeoff_train_size = 1000
mixed_training_shallow_cls = run_mixed_training_experiment(
    fit_shallow_neural_classifier_with_hyperparameter_search, bakeoff_train_size=bakeoff_train_size)

Stopping after epoch 23. Validation score did not improve by tol=1e-05 for more than 10 epochs. Final error is 0.3483838327229023

Best params: {'hidden_activation': Tanh(), 'hidden_dim': 200}
Best score: 0.546
Assessment dataset 1
              precision    recall  f1-score   support

    negative      0.626     0.685     0.654       428
     neutral      0.291     0.162     0.208       229
    positive      0.646     0.736     0.688       444

    accuracy                          0.597      1101
   macro avg      0.521     0.528     0.517      1101
weighted avg      0.565     0.597     0.575      1101

Assessment dataset 2
              precision    recall  f1-score   support

    negative      0.438     0.453     0.445       320
     neutral      0.578     0.520     0.547       612
    positive      0.512     0.573     0.541       429

    accuracy                          0.521      1361
   macro avg      0.510     0.515     0.511      1361
weighted avg      0.525     0.521     0.521      1361

Mean of macro-F1 scores: 0.514


### BERT encoding [2 points]

We might hypothesize that encoding our examples with BERT will yield improvements over the GloVe averaging method explored in the previous question, since BERT implements a much more complex and data-driven function for this kind of combination. This question asks you to begin exploring this general hypothesis.

Your task: write a function `hf_cls_phi` that uses Hugging Face functionality to encode individual examples with BERT and returns the final output representation above the [CLS] token.

You are not required to evaluate this feature function, but it is easy to do so with `sst.experiment` and `vectorize=False` (since your feature function directly encodes every example as a vector). Your code should also be a natural basis for even more powerful approaches – for example, it might be even better to pool all the output states rather than using just the first output state. Another option is [fine-tuning](finetuning.ipynb).

In [ ]:
from transformers import BertModel, BertTokenizer
import vsm

In [ ]:

# Instantiate a Bert model and tokenizer based on `bert_weights_name`:
bert_weights_name = 'bert-base-uncased'
#### YOUR CODE HERE
bert_tokenizer = BertTokenizer.from_pretrained(bert_weights_name)
bert_model = BertModel.from_pretrained(bert_weights_name)


def hf_cls_phi(text):
    # Get the ids. `vsm.hf_encode` will help; be sure to
    # set `add_special_tokens=True`.
    ##### YOUR CODE HERE
    batch_ids = vsm.hf_encode(text, bert_tokenizer, add_special_tokens=True)  # (1, seq_len)


    # Get the BERT representations. `vsm.hf_represent` will help:
    ##### YOUR CODE HERE
    reps = vsm.hf_represent(batch_ids, bert_model, layer=-1)   # (1, seq_len, 768)  (hidden_dim=768)
    
    cls_rep = reps[:, 0, :].squeeze(0)  # (1, 768)
    # Index into `reps` to get the representation above [CLS].
    # The shape of `reps` should be (1, n, 768), where n is the
    # number of tokens. You need the 0th element of the 2nd dim:
    ##### YOUR CODE HERE


    # These conversions should ensure that you can work with the
    # representations flexibly. Feel free to change the variable
    # name:
    return cls_rep.cpu().numpy()  # (1, 768)

In [ ]:
def test_hf_cls_phi(func):
    rep = func("Just testing!")

    expected_shape = (768,)
    result_shape = rep.shape
    assert rep.shape == (768,), \
        "Expected shape {}; got {}".format(
        expected_shape, result_shape)

    # String conversion to avoid precision errors:
    expected_first_val = str(0.1709)
    result_first_val = "{0:.04f}".format(rep[0])

    assert expected_first_val == result_first_val, \
        ("Unexpected representation values. Expected the "
        "first value to be {}; got {}".format(
            expected_first_val, result_first_val))

In [ ]:
if 'IS_GRADESCOPE_ENV' not in os.environ:
    test_hf_cls_phi(hf_cls_phi)

Note: encoding all of SST-3 train (no subtrees) takes about 11 minutes on my 2015 iMac, CPU only (32GB).

### Your original system [3 points]

Your task is to develop an original model for the SST-3 problem and our new bakeoff dataset. There are many options. If you spend more than a few hours on this homework problem, you should consider letting it grow into your final project! Here are some relatively manageable ideas that you might try:

1. We didn't systematically evaluate the `bidirectional` option to the `TorchRNNClassifier`. Similarly, that model could be tweaked to allow multiple LSTM layers (at present there is only one), and you could try adding layers to the classifier portion of the model as well.

1. We've already glimpsed the power of rich initial word representations, and later in the course we'll see that smart initialization usually leads to a performance gain in NLP, so you could perhaps achieve a winning entry with a simple model that starts in a great place.

1. Our [practical introduction to contextual word representations](finetuning.ipynb) covers pretrained representations and interfaces that are likely to boost the performance of any system.

We want to emphasize that this needs to be an __original__ system. It doesn't suffice to download code from the Web, retrain, and submit. You can build on others' code, but you have to do something new and meaningful with it. See the course website for additional guidance on how original systems will be evaluated.

In the cell below, please provide a brief technical description of your original system, so that the teaching team can gain an understanding of what it does. This will help us to understand your code and analyze all the submissions to identify patterns and strategies.  We also ask that you report the best score your system got during development (your best average of macro-F1 scores), just to help us understand how systems performed overall.

In [ ]:
import os
import pandas as pd
import numpy as np
from collections import Counter

import sst
import utils
import vsm
import torch
import torch.nn as nn
from torch_rnn_classifier import TorchRNNClassifier
from torch_shallow_neural_classifier import TorchShallowNeuralClassifier
from torch_tree_nn import TorchTreeNN

from transformers import BertModel, BertTokenizer
from sklearn.model_selection import PredefinedSplit, StratifiedKFold

In [ ]:
# PLEASE MAKE SURE TO INCLUDE THE FOLLOWING BETWEEN THE START AND STOP COMMENTS:
#   1) Textual description of your system.
#   2) The code for your original system.
#   3) The score achieved by your system in place of MY_NUMBER.
#        With no other changes to that line.
#        You should report your score as a decimal value <=1.0
# PLEASE MAKE SURE NOT TO DELETE OR EDIT THE START AND STOP COMMENTS

# NOTE: MODULES, CODE AND DATASETS REQUIRED FOR YOUR ORIGINAL SYSTEM
# SHOULD BE ADDED BELOW THE 'IS_GRADESCOPE_ENV' CHECK CONDITION. DOING
# SO ABOVE THE CHECK MAY CAUSE THE AUTOGRADER TO FAIL.

# START COMMENT: Enter your system description in this cell.
# My peak score was: MY_NUMBER
if 'IS_GRADESCOPE_ENV' not in os.environ:
    pass

# STOP COMMENT: Please do not remove this comment.

In [ ]:
# Create a BERT model with classification layer on either pooler output,
# or [CLS] last hidden representation
class HfBertClassifierModel(nn.Module):
    def __init__(self, n_classes, weights_name='bert-base-cased', pool=True):
        super().__init__()
        self.n_classes = n_classes
        self.weights_name = weights_name
        self.bert = BertModel.from_pretrained(self.weights_name)
        self.bert.train()
        self.hidden_dim = self.bert.embeddings.word_embeddings.embedding_dim
        self.pool = pool
        self.classifier_layer = nn.Linear(
            self.hidden_dim, self.n_classes)

    def forward(self, indices, mask):
        reps = self.bert(
            indices, attention_mask=mask)
        if self.pool:  # apply a cls layer on pooler output  (B, 768)
            out = self.classifier_layer(reps.pooler_output)
        else:  # apply a cls layer on [CLS] last hidden representation  (B, 768)
            out = self.classifier_layer(reps.last_hidden_state[:, 0, :])
        return out

In [ ]:
# a classifier to handle with new graph (model) and dataset (with masking and tokens' indices)
class HfBertClassifier(TorchShallowNeuralClassifier):
    def __init__(self, weights_name, pool=True, *args, **kwargs):
        self.weights_name = weights_name
        self.pool = pool
        self.tokenizer = BertTokenizer.from_pretrained(self.weights_name)
        super().__init__(*args, **kwargs)
        self.params += ['weights_name', 'pool']

    def build_graph(self):
        return HfBertClassifierModel(self.n_classes_, self.weights_name, self.pool)

    def build_dataset(self, X, y=None):
        data = self.tokenizer.batch_encode_plus(
            X,
            max_length=None,
            add_special_tokens=True,
            padding='longest',
            return_attention_mask=True)
        indices = torch.tensor(data['input_ids'])
        mask = torch.tensor(data['attention_mask'])
        if y is None:
            dataset = torch.utils.data.TensorDataset(indices, mask)
        else:
            self.classes_ = sorted(set(y))
            self.n_classes_ = len(self.classes_)
            class2index = dict(zip(self.classes_, range(self.n_classes_)))
            y = [class2index[label] for label in y]
            y = torch.tensor(y)
            dataset = torch.utils.data.TensorDataset(indices, mask, y)
        return dataset

In [ ]:
def hf_cls_phi(text):
    batch_ids = vsm.hf_encode(text, bert_tokenizer, add_special_tokens=True)  # (1, seq_len)
    reps = vsm.hf_represent(batch_ids, bert_model, layer=-1)   # (1, seq_len, 768)  (hidden_dim=768)
    cls_rep = reps[:, 0, :].squeeze(0)
    return cls_rep.cpu().numpy()  # (1, 768)

def bert_fine_tune_phi(text):
    return text

In [ ]:
def create_extended_train(train_df_subtree, max_examples=10000):
    """ create a balanced training set using subtrees

    Parameters
    ----------
    train_df_subtree : pd.DataFrame
        train reader where include_subtrees=True

    max_examples: int
        the amount of examples set for each label

    Returns
    -------
    pd.DataFrame
        a balanced train reader
    """
    tmp_train_df = train_df_subtree.loc[train_df_subtree['is_subtree'] == 0]
    label_dict = tmp_train_df['label'].value_counts().to_dict()
    dfs = []
    for label_name, label_count in label_dict.items():
        subtree_df = train_df_subtree.loc[
            (train_df_subtree['label'] == label_name) & (train_df_subtree['is_subtree'] == 1)]\
            .sample(MAX_EXAMPLES - label_count, random_state=1)
        dfs.append(subtree_df)

    train_df = pd.concat([tmp_train_df] + dfs, axis=0)
    del train_df_subtree
    return train_df

In [ ]:
def create_train_splitter(df, n_splits=5):
    """Create a balanced k fold, where subtrees must be on training set
    
    Parameters
    ----------
    df : pd.DataFrame
        training set dataframe

    n_splits: int
        the number of folds

    Returns
    -------
    sklearn.PredefinedSplit
        
    """
    index_df = df[['is_subtree', 'label']].rename(columns={"is_subtree": "index"})
    index_df['index'] = index_df['index'].apply(lambda val: -val)
    index_df['label'] = index_df['label'].replace({'positive': 1, 'negative': -1, 'neutral': 0})

    skf = StratifiedKFold(n_splits=n_splits)
    mask = index_df['index'] == 0
    for i, (train_index, test_index) in enumerate(skf.split(index_df.loc[mask, 'index'],
                                                            index_df.loc[mask, 'label'])):
        index_df.loc[index_df[mask].iloc[test_index].index, 'index'] = i


    train_splitter = PredefinedSplit(index_df['index'])
    return train_splitter


In [ ]:
# main- run experiments

MAX_EXAMPLES = 10000
bakeoff_train_size = 1000
n_splits = 5
DEVICE = 'cuda:0' if torch.cuda.is_available() else 'cpu'

SST_HOME = os.path.join('data', 'sentiment')

sst_dev = sst.dev_reader(SST_HOME)
bakeoff_dev = sst.bakeoff_dev_reader(SST_HOME)

full_train_df = sst.train_reader(SST_HOME, include_subtrees=True, dedup=True)
train_df = create_extended_train(full_train_df, max_examples=MAX_EXAMPLES)

train_df = pd.concat([train_df, bakeoff_dev.iloc[:bakeoff_train_size]], axis=0)
# reset index to avoid ambiguity in row index
train_df.reset_index(inplace=True, drop=True)

train_splitter = create_train_splitter(train_df, n_splits=n_splits)

assess_dfs = [sst_dev, bakeoff_dev.iloc[bakeoff_train_size:]]


DEVICE

'cuda:0'

In [ ]:
def fit_hf_bert_cls_classifier_with_hyperparameter_search(X, y):
    basemod = HfBertClassifier(
        device=DEVICE,
        weights_name='bert-base-cased',
        pool=False,
        batch_size=64,  # Small batches to avoid memory overload. (default: 8) change to 128? 1028?
        max_iter=100,  # We'll search based on 1 iteration for efficiency.
        n_iter_no_change=5,   # Early-stopping params are for the
        early_stopping=True)  # final evaluation.

    param_grid = {
        'gradient_accumulation_steps': [8],
        'eta': [0.0001],
        'hidden_dim': [300]}


    bestmod = utils.fit_classifier_with_hyperparameter_search(
        X, y, basemod, cv=train_splitter, param_grid=param_grid)

    return bestmod


def fit_hf_bert_pool_classifier_with_hyperparameter_search(X, y):
    basemod = HfBertClassifier(
        device=DEVICE,
        weights_name='bert-base-cased',
        pool=True,
        batch_size=64,  # Small batches to avoid memory overload. (default: 8)
        max_iter=100,  # We'll search based on 5 iteration for efficiency.  (default:  1)
        n_iter_no_change=5,   # Early-stopping params are for the
        early_stopping=True)  # final evaluation.

    param_grid = {
        'gradient_accumulation_steps': [8],
        'eta': [0.0001],
        'hidden_dim': [300]}


    bestmod = utils.fit_classifier_with_hyperparameter_search(
        X, y, basemod, cv=train_splitter, param_grid=param_grid)

    return bestmod

In [ ]:
def fit_hf_bert_cls(X, y):
    model = HfBertClassifier(
        device=DEVICE,
        weights_name='bert-base-cased',
        pool=False,
        gradient_accumulation_steps=8,
        eta=0.0001,
        hidden_dim=300,
        batch_size=64,  # Small batches to avoid memory overload. (default: 8) change to 128? 1028?
        max_iter=100,  # We'll search based on 1 iteration for efficiency.
        n_iter_no_change=5,   # Early-stopping params are for the
        early_stopping=True)  # final evaluation.
    model.fit(X, y)
    return model


def fit_hf_bert_pool(X, y):
    model = HfBertClassifier(
        device=DEVICE,
        weights_name='bert-base-cased',
        pool=True,
        gradient_accumulation_steps=8,
        eta=0.0001,
        hidden_dim=300,
        batch_size=64,  # Small batches to avoid memory overload. (default: 8)
        max_iter=100,  # We'll search based on 5 iteration for efficiency.  (default:  1)
        n_iter_no_change=5,   # Early-stopping params are for the
        early_stopping=True)  # final evaluation.

    model.fit(X, y)
    return model

In [ ]:
bert_weights_name = 'bert-base-uncased'
bert_tokenizer = BertTokenizer.from_pretrained(bert_weights_name)
bert_model = BertModel.from_pretrained(bert_weights_name)

In [ ]:
%%time
bert_cls_classifier_xval = sst.experiment(
    train_df,
    bert_fine_tune_phi,
    fit_hf_bert_cls,
    assess_dataframes=assess_dfs,
    vectorize=False)

Downloading:   0%|          | 0.00/208k [00:00<?, ?B/s]

Downloading:   0%|          | 0.00/29.0 [00:00<?, ?B/s]

Downloading:   0%|          | 0.00/426k [00:00<?, ?B/s]

Downloading:   0%|          | 0.00/570 [00:00<?, ?B/s]

Downloading:   0%|          | 0.00/416M [00:00<?, ?B/s]

Some weights of the model checkpoint at bert-base-cased were not used when initializing BertModel: ['cls.predictions.decoder.weight', 'cls.predictions.bias', 'cls.predictions.transform.LayerNorm.bias', 'cls.seq_relationship.bias', 'cls.predictions.transform.dense.bias', 'cls.predictions.transform.LayerNorm.weight', 'cls.seq_relationship.weight', 'cls.predictions.transform.dense.weight']
- This IS expected if you are initializing BertModel from the checkpoint of a model trained on another task or with another architecture (e.g. initializing a BertForSequenceClassification model from a BertForPreTraining model).
- This IS NOT expected if you are initializing BertModel from the checkpoint of a model that you expect to be exactly identical (initializing a BertForSequenceClassification model from a BertForSequenceClassification model).
Stopping after epoch 8. Validation score did not improve by tol=1e-05 for more than 5 epochs. Final error is 4.186468627420254

Assessment dataset 1
              precision    recall  f1-score   support

    negative      0.805     0.822     0.814       428
     neutral      0.532     0.183     0.273       229
    positive      0.711     0.937     0.809       444

    accuracy                          0.736      1101
   macro avg      0.683     0.648     0.632      1101
weighted avg      0.710     0.736     0.699      1101

Assessment dataset 2
              precision    recall  f1-score   support

    negative      0.680     0.625     0.651       320
     neutral      0.727     0.709     0.718       612
    positive      0.687     0.753     0.719       429

    accuracy                          0.703      1361
   macro avg      0.698     0.696     0.696      1361
weighted avg      0.703     0.703     0.703      1361

Mean of macro-F1 scores: 0.664
CPU times: user 1h 1min 34s, sys: 14min 7s, total: 1h 15min 42s
Wall time: 1h 15min 42s


In [ ]:
%%time
bert_pool_classifier_xval = sst.experiment(
    train_df,
    bert_fine_tune_phi,
    fit_hf_bert_pool,
    assess_dataframes=assess_dfs,
    vectorize=False) # Pass in the BERT hidden state directly!

Some weights of the model checkpoint at bert-base-cased were not used when initializing BertModel: ['cls.predictions.decoder.weight', 'cls.predictions.bias', 'cls.predictions.transform.LayerNorm.bias', 'cls.seq_relationship.bias', 'cls.predictions.transform.dense.bias', 'cls.predictions.transform.LayerNorm.weight', 'cls.seq_relationship.weight', 'cls.predictions.transform.dense.weight']
- This IS expected if you are initializing BertModel from the checkpoint of a model trained on another task or with another architecture (e.g. initializing a BertForSequenceClassification model from a BertForPreTraining model).
- This IS NOT expected if you are initializing BertModel from the checkpoint of a model that you expect to be exactly identical (initializing a BertForSequenceClassification model from a BertForSequenceClassification model).
Stopping after epoch 10. Validation score did not improve by tol=1e-05 for more than 5 epochs. Final error is 3.2579322119709104

Assessment dataset 1
              precision    recall  f1-score   support

    negative      0.828     0.776     0.801       428
     neutral      0.412     0.266     0.324       229
    positive      0.725     0.901     0.803       444

    accuracy                          0.720      1101
   macro avg      0.655     0.648     0.643      1101
weighted avg      0.700     0.720     0.703      1101

Assessment dataset 2
              precision    recall  f1-score   support

    negative      0.671     0.631     0.651       320
     neutral      0.743     0.765     0.754       612
    positive      0.744     0.746     0.745       429

    accuracy                          0.727      1361
   macro avg      0.719     0.714     0.716      1361
weighted avg      0.726     0.727     0.727      1361

Mean of macro-F1 scores: 0.680
CPU times: user 1h 17min 3s, sys: 17min 51s, total: 1h 34min 55s
Wall time: 1h 34min 45s


In [ ]:
# McNemar test

def mcnemar_test(exp1, exp2, num_assess):
    """
    McNemar's test using the chi2 distribution.

    Parameters
    ----------
    exp1 : dict
        experiment1 output

    exp2 : lists
        experiment2 output

    num_assess: int
        Run McNemar's test on assessment dataset number "num_assess".
        If num_assess = -1, then run test on all datasets 

    Returns
    -------
    float, float (the test statistic and p value)
    """
    total_assess = len(exp1['assess_datasets'])
    assert num_assess < total_assess, \
            "num_asses ({}) is larger than total datasets in assessments ({})".format(
            num_assess, total_assess)
    
    if num_assess >= 0:  # apply test on specific dataset
        m = utils.mcnemar(
            exp1['assess_datasets'][num_assess]['y'],  # list
            exp1['predictions'][num_assess],  # ndarray
            exp2['predictions'][num_assess]  # ndarray
            )
    else:  # apply test on all datasets
        m = utils.mcnemar(
            [exp1['assess_datasets'][i]['y'] for i in range(total_assess)],  # list
            np.concatenate([exp1['predictions'][i] for i in range(total_assess)]),  # ndarray
            np.concatenate([exp2['predictions'][i] for i in range(total_assess)]),  # ndarray
            )   
    return m

In [ ]:
m0 = mcnemar_test(bert_cls_classifier_xval, bert_pool_classifier_xval, num_assess=0)
m0

(2.048, 0.15240628395674527)

In [ ]:
m1 = mcnemar_test(bert_cls_classifier_xval, bert_pool_classifier_xval, num_assess=1)
m1

(5.361256544502618, 0.02058893364168301)

In [ ]:
m_all = mcnemar_test(bert_cls_classifier_xval, bert_pool_classifier_xval, num_assess=-1)
m_all

/content/drive/My Drive/NLU/HW/HW2/utils.py:158: RuntimeWarning: divide by zero encountered in double_scalars
  stat = ((np.abs(c10 - c01) - 1.0)**2) / (c10 + c01)


(inf, 0.0)

## Bakeoff [1 point]

As we said above, the bakeoff evaluation data is the official SST test-set release and a new test set derived from the same sources and labeling methods as for `bakeoff_dev`.

For this bakeoff, you'll evaluate your original system from the above homework problem on these test sets. Our metric will be the mean of the macro-F1 values, which weights both datasets equally despite their differing sizes.

The central requirement for your system is that you have define a `predict_one` method for it that maps a text (str) directly to a label prediction – one of 'positive', 'negative', 'neutral'. If you used `sst.experiment` with `vectorize=True`, then the following function (for `softmax_experiment`) will be easy to adapt – you probably just need to change the variable `softmax_experiment` to the variable for your experiment output.

In [ ]:
def predict_one_softmax(text):
    # Singleton list of feature dicts:
    feats = [softmax_experiment['phi'](text)]
    # Vectorize to get a feature matrix:
    X = softmax_experiment['train_dataset']['vectorizer'].transform(feats)
    # Standard sklearn `predict` step:
    preds = softmax_experiment['model'].predict(X)
    # Be sure to return the only member of the predictions,
    # rather than the singleton list:
    return preds[0]

If you used an RNN like the one we demoed above, then featurization is a bit more straightforward:

In [ ]:
def predict_one_rnn(text):
    # List of tokenized examples:
    X = [rnn_experiment['phi'](text)]
    # Standard `predict` step on a list of lists of str:
    preds = rnn_experiment['model'].predict(X)
    # Be sure to return the only member of the predictions,
    # rather than the singleton list:
    return preds[0]

The following function is used to create the bakeoff submission file. Its arguments are your `predict_one` function and an output filename (str).

In [ ]:
def create_bakeoff_submission(
        predict_one_func,
        output_filename='cs224u-sentiment-bakeoff-entry.csv'):

    bakeoff_test = sst.bakeoff_test_reader(SST_HOME)
    sst_test = sst.test_reader(SST_HOME)
    bakeoff_test['dataset'] = 'bakeoff'
    sst_test['dataset'] = 'sst3'
    df = pd.concat((bakeoff_test, sst_test))

    df['prediction'] = df['sentence'].apply(predict_one_func)

    df.to_csv(output_filename, index=None)

Thus, for example, the following will create a bake-off entry based on `predict_one_softmax`:

In [ ]:
# This check ensure that the following code only runs on the local environment only.
# The following call will not be run on the autograder environment.
if 'IS_GRADESCOPE_ENV' not in os.environ:
    pass
    create_bakeoff_submission(predict_one_softmax)

This creates a file `cs224u-sentiment-bakeoff-entry.csv` in the current directory. That file should be uploaded as-is. Please do not change its name.

Only one upload per team is permitted, and you should do no tuning of your system based on what you see in our bakeoff prediction file – you should not study that file in anyway, beyond perhaps checking that it contains what you expected it to contain. The upload function will do some additional checking to ensure that your file is well-formed.

People who enter will receive the additional homework point, and people whose systems achieve the top score will receive an additional 0.5 points. We will test the top-performing systems ourselves, and only systems for which we can reproduce the reported results will win the extra 0.5 points.

Late entries will be accepted, but they cannot earn the extra 0.5 points.

## Submission Instruction

Submit the following files to gradescope submission

- Please do not change the file name as described below
- `hw_sentiment.ipynb` (this notebook)
- `cs224u-sentiment-bakeoff-entry.csv` (bake-off output)


In [ ]:
# make prediction with best BERT model
def predict_one_bert(text):
    # List of tokenized examples:
    X = [bert_pool_classifier_xval['phi'](text)]
    # Standard `predict` step on a list of lists of str:
    preds = bert_pool_classifier_xval['model'].predict(X)
    # Be sure to return the only member of the predictions,
    # rather than the singleton list:
    return preds[0]

In [ ]:
if 'IS_GRADESCOPE_ENV' not in os.environ:
    create_bakeoff_submission(predict_one_bert)